gobblin-modules/gobblin-kafka-08/src/main/java/org/apache/gobblin/metrics/kafka/KafkaKeyValueProducerPusher.java

```diff
@@ -80,17 +88,46 @@ public KafkaKeyValueProducerPusher(String brokers, String topic) {
    */
   public void pushMessages(List<Pair<K, V>> messages) {
     for (Pair<K, V> message: messages) {
-      this.producer.send(new ProducerRecord<>(topic, message.getKey(), message.getValue()), (recordMetadata, e) -> {
+      this.futures.offer(this.producer.send(new ProducerRecord<>(topic, message.getKey(), message.getValue()), (recordMetadata, e) -> {
         if (e != null) {
           log.error("Failed to send message to topic {} due to exception: ", topic, e);
         }
-      });
+      }));
+    }
+
+    //Accumulate futures returned from send() into a buffer; will be used to simulate flush by calling get() on
+    // each of the accumulated futures.
+    if (this.futures.size() >= MAX_NUM_FUTURES_TO_BUFFER) {
+      flush(MAX_NUM_FUTURES_TO_BUFFER);
+      this.futures.clear();
```

Comment:
This won't be thread safe since if there are concurrent calls to pushMessages() some unacknowledged futures may be cleared. `flush` is already removing with `poll`, so I think this can be removed.

Also, this forces a synchronous wait every 1000 messages. Should there be different max buffer and flush amounts to avoid waiting for the newest messages?

Reply:
Good catch! Yes, the clear() is not needed. Also, changed the pushMessages() method to exclude the newest added messages.